### PR TITLE
#149 : Add list command for dicedb

### PR DESCRIPTION
**Pull request by yashs360**
Added list command. Please review and provide your inputs.

Summary of change:
- Added Constants for Command names to keep it easy to add new commands, as well as reuse commands in tests
- Added List command
- Added test and benchmark

Benchmark results - (first time running benchmark so i don't have a baseline to compare if this code was good or shit). I could use some help here.

Screenshot 2024-07-12 at 11 12 24 PM

### CODE REVIEW

**Comment by yashs360** [comment]
@arpitbbhayani @JyotinderSingh - First real code contribution, not like the past fluffy PRs. Please review.

File: core/eval.go [hunk 1]
```diff
@@ -844,93 +897,101 @@ func evalQWATCH(args []string, c *Client) []byte {
 	return RESP_OK
 }
 
+// evalLIST returns an array of commands supported by DiceDB
+func evalLIST() []byte {
+	sort.Strings(diceCommands)
+	return Encode(strings.Join(diceCommands, ", "), false)
+}
+
 func executeCommand(cmd *RedisCmd, c *Client) []byte {
```
Comment (yashs360) [inline core/eval.go:906]: Everything below this in this file is mostly cleanup to use the constant command names.

File: core/eval.go [hunk 2]
```diff
@@ -19,10 +21,61 @@ var RESP_MINUS_1 []byte = []byte(":-1\r\n")
 var RESP_MINUS_2 []byte = []byte(":-2\r\n")
 var RESP_EMPTY_ARRAY []byte = []byte("*0\r\n")
 
+// Allsupported Dice Commands
```
Comment (JyotinderSingh) [inline core/eval.go:24]: nit: *All supported

File: tests/commands_test.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+		"DEL", "DISCARD", "EXEC", "EXPIRE", "GET", "INCR", "INFO", "LATENCY", "LIST", "LRU",
+		"MULTI", "PING", "QINTINS", "QINTLEN", "QINTPEEK", "QINTREM", "QREFINS", "QREFLEN",
+		"QREFPEEK", "QREFREM", "QWATCH", "SET", "SLEEP", "STACKINTLEN", "STACKINTPEEK",
+		"STACKINTPOP", "STACKINTPUSH", "STACKREFLEN", "STACKREFPEEK", "STACKREFPOP",
+		"STACKREFPUSH", "SUBSCRIBE", "TTL",
+	}
+
+	subscriber := getLocalConnection()
+
+	rp := fireCommandAndGetRESPParser(subscriber, "LIST")
```
Comment (JyotinderSingh) [inline tests/commands_test.go:23]: No need to use `fireCommandAndGetRESPParser` here, you can directly use the `fireCommand` utility.

`fireCommandAndGetRESPParser` is used for commands that may return more data over time (eg. QWATCH)

File: tests/commands_test.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+	}
+}
+
+func BenchmarkListCommand20(t *testing.B) {
+	call(20, t)
+}
+
+func BenchmarkListCommand200(t *testing.B) {
+	call(200, t)
+}
```
Comment (JyotinderSingh) [inline tests/commands_test.go:54]: I would imagine 20 and 200 are small and close enough for their benchmarks to not be very different. You could probably remove one of them.

File: tests/commands_test.go [hunk 5]
```diff
@@ -0,0 +1,58 @@
+package tests
```
Comment (JyotinderSingh) [inline tests/commands_test.go:1]: nit: This file should probably be called `list_test.go`?

**Review by JyotinderSingh — COMMENTED** [review]
Left some minor comments.

File: tests/commands_test.go (hunk 5 above)
Comment (yashs360) [reply tests/commands_test.go:1]: I was split mind here. We have 39 commands and i was thinking do we create 39 separate tests or do one test for all the commands. I'll rename.

File: tests/commands_test.go (hunk 3 above)
Comment (yashs360) [reply tests/commands_test.go:23]: Interesting. Thanks

**Review by JyotinderSingh — APPROVED** [review]

**Comment by JyotinderSingh** [comment]
Thanks for this contribution!

**Comment by JyotinderSingh** [comment]
I re-read the associated issue, turns out the actual command name needs to be `COMMAND LIST`, and we should also include a memory benchmark.
Let's do that before we merge this.

**Pull request merged**